File: 프로그래머스/JavaScript/Lv. 2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
  let cnt = 0;
  for (let i = 0; i < s.length; i++) {
    let isTrue = true;
    const stack = [];

    for (let j = 0; j < s.length; j++) {
      const n = s[(i + j) % s.length];

      if (n == '(' || n == '{' || n == '[') {
        stack.push(n);
      } else {
        if (stack.length == 0) {
          isTrue = false;
          break;
        }

        const top = stack[stack.length - 1];
        if (top == '{' && n == '}') stack.pop();
        else if (top == '(' && n == ')') stack.pop();
        else if (top == '[' && n == ']') stack.pop();
        else {
          isTrue = false;
          break;
        }
      }
    }

    if (isTrue == true && stack.length == 0) {
      cnt += 1;
    }
  }
  return cnt;
}

File: 프로그래머스/Java/Lv. 2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.ArrayDeque;

class Solution {
	public int solution(String s) {
		int answer = 0;
		int n = s.length();

		for (int i = 0; i < n; i++) {
			ArrayDeque<Character> stack = new ArrayDeque<>();
			boolean isValid = true;

			for (int j = 0; j < n; j++) {
				char c = s.charAt((i + j) % n);

				if (!stack.isEmpty() && ((stack.peek() == '(' && c == ')') || (stack.peek() == '{' && c == '}')
						|| (stack.peek() == '[' && c == ']'))) {
					stack.pop();
				} else if (c == ')' || c == '}' || c == ']') {
					isValid = false;
					break;
				} else {
					stack.push(c);
				}
			}

			if (isValid && stack.isEmpty()) {
				answer++;
			}
		}

		return answer;
	}
}